pkgs/development/haskell-modules/make-package-set.nix

```diff
@@ -257,20 +261,46 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
     shellFor = { packages, withHoogle ? false, ... } @ args:
       let
         selected = packages self;
-        packageInputs = builtins.map getBuildInputs selected;
-        haskellInputs =
-          builtins.filter
-            (input: pkgs.lib.all (p: input.outPath != p.outPath) selected)
-            (pkgs.lib.concatMap (p: p.haskellBuildInputs) packageInputs);
+
+        packageInputs = map getBuildInputs selected;
+
+        # If `packages = [ a b ]` and `a` depends on `b`, don't build `b`, such
+        # that `a` will take the `b` dependency from the cabal.project file.
+        # This allows changing `b` while in the shell and have `a` use the
+        # changes directly.
```

Comment:
`cabal new-build` is happy to ignore any installed `b` actually. The real reason for this is to avoid building `b` with Nix at all, so that entering the nix shell doesn't require rebuilding `b` every time it changes when you're just going to ignore that version of `b`.

Reply:
Ahh good point, will adjust that message